### PR TITLE
Add API to tell if user code is in active IRQn

### PR DESCRIPTION
**Pull request by AlessandroA**
This introduces a new user API:

`int vIRQ_GetLevel(void)`

that returns:
- `-1`  if no interrupt is active
- `>= 0` if an interrupt is active and it is being served;
       the actual return value indicates the priority level of the IRQn
       as set by default or with `vIRQ_SetPriority(IRQn_Type irqn)`

### CODE REVIEW

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by AlessandroA** [comment]
@meriac 
@bremoran please tell me if this implements what you need, and if it solves https://github.com/ARMmbed/uvisor/issues/34

**Comment by AlessandroA** [comment]
@hugovincent adding you as you were involved in the original discussion. Feel free to comment

File: core/mbed/source/interrupts.cpp [hunk 1]
```diff
@@ -106,3 +106,38 @@ uint32_t vIRQ_GetPriority(uint32_t irqn)
...
+ * a preprocessor define in a uniform way it will be deleted from here */
+#define UVISOR_NVIC_IRQn_OFFSET 16
+
+int vIRQ_GetLevel(void)
+{
+    /* if uVisor is disabled we use the standard priority levels */
+    if (__uvisor_mode == 0) {
+        /* check if an IRQn is active (an ISR is being served) */
+        uint32_t ipsr = __get_IPSR();
+        int irqn = (int) (ipsr & 0x1FF) - UVISOR_NVIC_IRQn_OFFSET;
```
Comment (meriac) [inline core/mbed/source/interrupts.cpp]: Ensure that mask is appropriate for selected target. The NVIC->IABR register (as use in NVIC_GetActive) is only 8 DWORD wide - which leaves only 256 interrupts. 0x1FF would reult in a overflow.

File: core/system/src/unvic.c [hunk 2]
```diff
@@ -196,6 +196,31 @@ uint32_t unvic_irq_priority_get(uint32_t irqn)
...
 
+int unvic_irq_level_get(void)
+{
+    /* get and validate the unprivileged exception stack frame */
+    uint32_t *box_sp = svc_cx_validate_sf((uint32_t *) __get_PSP());
+
+    /* gather IPSR from exception stack frame */
+    /* the currently active IRQn is the one of the SVCall, while instead we want
+     * to know the IRQn at the time of the SVCcall, which is saved on the stack */
+    uint32_t ipsr = box_sp[7];
```
Comment (meriac) [inline core/system/src/unvic.c]: please drop svc_cx_validate_sf and use unprivileged read instead

File: core/system/src/unvic.c [hunk 3]
```diff
@@ -196,6 +196,31 @@ uint32_t unvic_irq_priority_get(uint32_t irqn)
...
+    uint32_t irqn = (ipsr & 0x1FF) - IRQn_OFFSET;
```
Comment (meriac) [inline core/system/src/unvic.c]: 0x1FF results in an overflow of NVIC_GetActive

**Comment by miklis** [comment]
Build finished. No test results found.

**Comment by AlessandroA** [comment]
@meriac 

- removed use of `svc_cx_validate_sf()`
- added check on `irqn` (fail if `>= HW_INTERRUPTS`)
    - note: this check is included in `unvic_check_default()` in uVisor, and is added explicitly in the mbed-facing fallback version (for when uVisor is disabled)

**Pull request merged**